### ISSUE TITLE
Making xopen more suitable for use with type annotation.

### ISSUE THREAD

**Issue by rhpvorderman** [pycompression/xopen#22]
Currently if I use 
```python
with open('my_file', 'rb') as file_handler:
    pass
``` 
I know that file_handler is of type `io.BufferedReader`. This allows me to easily find out which methods it has.

With xopen:

```python
with xopen('my_file', 'rb') as file_handler:
    pass
``` 

This can be any of several, including `io.BufferedReader`, `xopen.PipedGzipReader`, `lzma.LzmaFile` etc. These have some methods which are the same and some which are different.

It would be nice in the future to have four abstract base classes xopen. (Binary/Text)(Reader/Writer). So that when xopen is used you can type the resulting output file and know its methods.

This is not going to be trivial to implement but I would like to throw this idea out there. It will make coding easier for the 32 projects that are currently using it downstream. It is something on the wishlist after python 2 support is definitely dropped.

**Comment by marcelm** [comment]
Do you mean to add type annotation suitable for type checkers and IDEs? I’d support that.

Note that the number of downstream projects using xopen is lower. Many of those listed by GitHub on the 'Used by' page are false positives, or use xopen indirectly through Cutadapt.

**Comment by marcelm** [comment]
I think this was solved by f6c04d7
